### ISSUE TITLE
RFC: syntax for efficient await abstraction

### ISSUE THREAD

**Issue by crusso** [dfinity/motoko#3602]
Consider the following async calculator code. It's ridiculously inefficient due to the recursive awaits.

``` motoko
actor a {

  public /* query */ func pow(i :Int, n:Int) : async Int {
     power(i, n);
  };

  public type Expression = {
    #const : Int;
    #add : (Expression, Expression);
    #pow : (Expression, Expression);
  };

  // slow version using async
  func eval(exp : Expression) : async Int {
//  func eval(exp : Expression) : async Int = async {
    switch (exp) {
      case (#const(n)) n;
      case (#add(e1, e2)) (await eval(e1)) + (await eval(e2));
      case (#pow(e1, e2)) await (pow(await (eval e1), await (eval e2)));
    }
  };
}
```

https://github.com/dfinity/motoko/pull/3573 fixes the inefficiency, but only by making some await's commit, others not, which Andreas objected too. Perhaps rightly.

To fix that, I'm proposing a new type of async computation (a thunk), together with new introduction and elim forms, to introduce a clear distinction between fast-and-loose and slow-but-safe async abstraction.

The one catch is that the new type is a monadic *value* and won't have any effect unless `run `with `await in` (`await?`) which opens the footgun that someone creates the thunk but doesn't eliminate it. I could try to avoid that by warning about free-standing, unawaited `await T` (or `async? T`) values or making the type second class somehow (like our mutable variables).
Of course, the question now is, what syntax for the type constructor and terms:

## Option 1 (👍)
type: `await T`
intro: `do await <block>`
elim: `await in <exp>`  (note the re-use of `in`)
``` motoko
  // Use `await Int`/`do await {}`/`await in ..` to avoid context switch at each recursive call
  func eval(exp : Expression) : await Int {
//  func eval(exp : Expression) : await Int  = do await { ...}
    switch (exp) {
      case (#const(n)) n;
      case (#add(e1, e2)) (await in eval(e1)) + (await in eval(e2));
      case (#pow(e1, e2)) await pow(await in (eval e1), await in (eval e2));
                              // ^^^ only real context switch on call to asynchronous `pow` query
    }
  };
```
+ backwards compatible
-  verbose

## Option 2 (👎)

type: `async? T`
intro: `async? <block>`
elem: `await? <exp>`

``` motoko
  // Use `async? Int`/`async? {}`/`await?` to avoid context switch at each recursive call
  func eval(exp : Expression) : async? Int {
//  func eval(exp : Expression) : async? Int = async? { ... }
    switch (exp) {
      case (#const(n)) n;
      case (#add(e1, e2)) (await? eval(e1)) + (await? eval(e2));
      case (#pow(e1, e2)) await pow(await? (eval e1), await? (eval e2));
                        // ^^^ only real context switch on call to asynchronous `pow` query
    }
  };
```
+ minor variation of existing syntax
- might not fit the grammar because of options (need to check)
- footgun is the different behaviour of async vs async?

## Option 3 (:smile:)

type: `thunk T`
intro: `thunk <block>`
elim: `force <exp>`

Seems to be closest to the essential ideas, but burns keywords and is a breaking change.

## Option 4 (:tada:)

type: `async* T`
intro: `async* <block>`
elim: `await* <exp>`

+ avoids confusion with option, probably backwards compatible.
+ minor variation of existing syntax
- footgun is the different behaviour of async vs async*

**Comment by crusso** [comment]
# Option 4

`async*` and `await*` to avoid confusion with options.

**Comment by nomeata** [comment]
I immediatelly liked `await in …`, only to then also immediately like `async?` and `await?`. So I my favorite is Option 2 so far.

**Comment by ggreif** [comment]
Since these are first-class values, the monadic Option 2 looks ok to me. If it doesn't work out grammar-wise, then option 4 would be the runner up.

**Comment by ggreif** [comment]
Since this all is happening under tight temporal constraints, maybe we shouldn't advertise this feature as a stable one. Softening the narrative would give us some leeway changing the syntax if we come up with a better one (deprecation the previous).

**Comment by ggreif** [comment]
Another thing:
- `async? T` is non-sharable,
- so only available in private function signatures

Right?

**Comment by nomeata** [comment]
> Since this all is happening under tight temporal constraints

Interesting how quickly the constraints can change, given that we have been discussing this per years when nobody seemed to care… oh well, I guess that's Realpolitik

**Comment by ggreif** [comment]
I brought this up with Andreas at ICFP too, it was well-known, but we always kept discussing the necessity and semantics, never the _syntax_ (other than `do not await` :-)
Now a customer feels the need. Realpolitik indeed.

**Comment by rossberg** [comment]
What's the emoji to pick for "none of these"? : )

So yeah, FTR, I haven't changed my mind. I still think this is a very bad idea that causes way more damage than good. It makes an already difficult atomicity/commit/rollback story even less understandable and way less predictable. It creates a serious, seductive footgun, where users for sure will prematurely optimise their code (because fast) with subtle concurrency bugs, ruining all benefits of the actor model.

What is the advice to users? When is it safe to use this as opposed to proper async/await? I honestly have no idea. The only obvious case is where all code involved (including remote calls) is known to be stateless. But the provides no useful support or guidance regarding state dependencies, which I think is a much more pressing problem that would need solving first before going fast and loose with commits and atomicity. ; )

**Comment by crusso** [comment]
What exactly are you objecting to @rossberg? 

I'm suggesting introducing a type distinction to address your original concerns  - the old await will remain definite a commit point and yield/commit on both pending and completed futures.

Real users have actually complained about the cost of abstraction and the pressure on the (limited) self-call queue that the extra yields entail, much as we predicted all along.

(At this point, I'm tempted to go with the original solution which was at least elegant if a mild foot-gun. It was also opt-in but arguably in a more subtle way - I was hoping the type distinction would alleviate your concerns.)

**Comment by crusso** [comment]
Here's the related forum post 
https://forum.dfinity.org/t/canister-output-message-queue-limits-and-ic-management-canister-throttling-limits/15972/9

In particular this one:

https://forum.dfinity.org/t/canister-output-message-queue-limits-and-ic-management-canister-throttling-limits/15972/16?u=claudio

I was originally hoping that the replica optimization of allowing multiple messages to self in the same round would reduce the latency of async expressions, but, given the artificial (but probably unavoidable) queue limits imposed by the IC, that only gets you so far.

Plus its very reasonable to want abstraction without breaking atomicity every time you use it, regardless of what the callee is doing.

**Comment by crusso** [comment]
> Another thing:
> 
> * `async? T` is non-sharable,
> * so only available in private function signatures
> 
> Right?

Yes, it's non-shareable, and I'd like to lift the restriction that T must be shared (as we should for ordinary async too).

**Comment by rossberg** [comment]
@crusso, my concern hasn't changed: when I see `await* f()`, I _cannot tell_ whether it allows interleaving and worse, whether it results in a commit. It may or may not, depending on whatever the F happens deep inside `f` on a particular execution. That's terrible from a correctness and modular reasoning point of view. If there ever is a rollback, you have approximately zero chance to predict or control the outcome and the overall consistency of the app's state. Concurrency is already hairy enough without such uncertainty. In the context of the IC, this isn't just a harmless thunk, nor a mere performance shortcut.

Wrt user requests, I still think the alleged Henry Ford quote about faster horses applies. Deeply nested async is bad code, I think, just like putting all your Haskell functions inside the IO monad. Do not "fix" bad code by language hacks, make the language encourage better code.

But I'll shut up, it's up to you to decide.

**Comment by crusso** [comment]
When you see `await*` you now know that the code may yield and commit state changes, so those changes better be complete and ready for publication to other concurrent observers, or private and irrelevant to others if rolled back. 

At the moment, Motoko actively encourages worse code, by forcing users to avoid abstraction, maintain their own stacks, and even count issued self calls.

**Comment by ggreif** [comment]
Yep. The tool named `await*` won't take away the classical `await` from the user's toolbox. It just gives you another tool with different characteristics, advantages as well as drawbacks. It makes certain things possible that weren't before (Like `MonadFix` vs. `Monad` in Haskell).

**Comment by crusso** [comment]
> I immediatelly liked `await in …`, only to then also immediately like `async?` and `await?`. So I my favorite is Option 2 so far.

(sorry, edited your comment by mistake, how rude!)

Bugger: `async?Int` would be a breaking change since it already parses as `async ?Int` and would now parse as `async? Int`, changing semantics. I guess we could detect and warn about the change.

`async*`,`await*` doesn't have this problem and still suggests zero or more commits/suspensions which is kinda more accurate.

**Comment by crusso** [comment]
ASYNC/AWAIT just to shout ;->

**Comment by rossberg** [comment]
@crusso:
> When you see `await*` you now know that the code may yield and commit state changes,

Emphasis on "may", which makes this observationally equivalent to knowing nothing. ; )

> so those changes better be complete and ready for publication to other concurrent observers, or private and irrelevant to others if rolled back.

That isn't sufficient, though. You no longer know _how far_ the rollback will reach. Consequently, any changes before that point must _simultaneously_ be safe to commit and safe _not_ to commit, regardless of what happens after. I suppose I lack imagination for what sort of changes would fall into that category, and how programmers would get this right. In practice, I suspect its only safe when it's pure or they have a proper commit point before it, but that of course would defeat the purpose.

> At the moment, Motoko actively encourages worse code, by forcing users to avoid abstraction, maintain their own stacks, and even count issued self calls.

I guess I beg to differ about which code is worse. I thought there is wide consensus that I/O deeply nested in program logic is a particularly bad anti-pattern. And remote calls are the IC's version of I/O!

@ggreif:
> Yep. The tool named `await*` won't take away the classical `await` from the user's toolbox.

Except that the average dev will happily pick `await*` over `await`, because fast. Especially when they do not really understand the difference, which I'd expect to be the common case.

**Comment by crusso** [comment]
Given the syntactic problems with `async?`/`await?`, and my reluctance to introduce new keywords that were identifiers, I suggest we go with `async*` and `await*`. 

Object now or forever hold your peace..

**Comment by crusso** [comment]
> @crusso:
> 
> > When you see `await*` you now know that the code may yield and commit state changes,
> 
> Emphasis on "may", which makes this observationally equivalent to knowing nothing. ; )
> 
> > so those changes better be complete and ready for publication to other concurrent observers, or private and irrelevant to others if rolled back.
> 
> That isn't sufficient, though. You no longer know _how far_ the rollback will reach. Consequently, any changes before that point must _simultaneously_ be safe to commit and safe _not_ to commit, regardless of what happens after. I suppose I lack imagination for what sort of changes would fall into that category, and how programmers would get this right. In practice, I suspect its only safe when it's pure or they have a proper commit point before it, but that of course would defeat the purpose.

This is actually a very common pattern in the implementation of lock free data structures, where you build up a bunch of private state off to the side until you reach a final, definite commit point. Commiting or rolling back local changes that only affect your own logical state isn't problematic as long as you haven't published the changes to other observers.


> > At the moment, Motoko actively encourages worse code, by forcing users to avoid abstraction, maintain their own stacks, and even count issued self calls.
> 
> I guess I beg to differ about which code is worse. I thought there is wide consensus that I/O deeply nested in program logic is a particularly bad anti-pattern. And remote calls are the IC's version of I/O!

Right, but this will at least be marked with async*/await*, so not totally hidden from the user.

> @ggreif:
> 
> > Yep. The tool named `await*` won't take away the classical `await` from the user's toolbox.
> 
> Except that the average dev will happily pick `await*` over `await`, because fast. Especially when they do not really understand the difference, which I'd expect to be the common case.

Well, it's not like async/await doesn't introduces it own slew of concurrency hazards already. So yes, it's a danger but I expect our programmers are already not aware that a local await causes a context switch. Indeed, many were surprised that it did, opening them up to concurrency bugs.

**Comment by rossberg** [comment]
> This is actually a very common pattern in the implementation of lock free data structures

Okay, but those are as expert-level as it can possibly get, nothing to leave in the hands of mere mortals.

> So yes, it's a danger but I expect our programmers are already not aware that a local await causes a context switch. Indeed, many were surprised that it did, opening them up to concurrency bugs.

Hm, so, because the language already is a bit confusing and dangerous, it's fine to make it more confusing and dangerous? I'm not sure... ;-)

**Comment by crusso** [comment]
Objection noted. 
PR is here #3609

**Comment by crusso** [comment]
Closed with #3609
